### PR TITLE
feat(stream): add support for xreadgroup command

### LINKED ISSUE #1433: Implement `XREADGROUP` command
Command reference info - https://redis.io/commands/xreadgroup/

### PR DESCRIPTION
**Pull request by Abhra303**
XREADGROUP lets consumers read entries from streams. This PR implements the `XREADGROUP` command.

Fixes #1433

### CODE REVIEW

**Review by dranikpg — CHANGES_REQUESTED** [review]
1. Please refactor ParseReadArgsOrReply, its 200+ loc and does too much
2. When refactoring, please make the high level logic easy to be understood, no hidden hops or schedules
3. You can't reply from a transaction callback, you should atomically store errors and handle them in the command function

File: src/server/stream_family.cc
Comment (kostasrim) [inline src/server/stream_family.cc:1891]: These two functions only differ on the boolean flag `true/false` passed down to `ParseReadArgsOrReply`, extract the common logic into a separate function and call it from there (see suggestion below).

Comment (kostasrim) [inline src/server/stream_family.cc:1891]:
```suggestion
void SomeGoodNameHere(CmdArgList args, bool val, ConnectionContext* cntx) {
  auto opts = ParseReadArgsOrReply(args, val, cntx);
  if (!opts) {
    return;
  }
  return XReadGeneric(args, opts, cntx);
}

void StreamFamily::XRead(CmdArgList args, ConnectionContext* cntx) {
  return SomeGoodNameHere(args, false, cntx);
}

void StreamFamily::XReadGroup(CmdArgList args, ConnectionContext* cntx) {
  return SomeGoodNameHere(args, true, cntx);
}

```

File: src/server/stream_family.cc
Comment (kostasrim) [inline src/server/stream_family.cc:718]: Too verbose, consider adding a constructor (I would suggest the non standard before C++ 20 https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers but I really don't like non conforming C++ standard code -- well with a few exceptions ;))

File: src/server/stream_family.cc
Comment (kostasrim) [inline src/server/stream_family.cc:1573]: I would love for C++ to allow `[std::ignore, is_insterted]` in structure bindings but yeah, they didn't like the symmetry with `std::tie`. Ignore my comment, just a small rant.

File: src/server/stream_family.cc
Comment (kostasrim) [inline src/server/stream_family.cc:1572]: `list init` or consider adding a constructor

File: src/server/stream_family.cc
Comment (kostasrim) [reply src/server/stream_family.cc:1589]: goto is really bug prone. I will check it out explicitly and see if we can have a quick workaround and come back to you here. Otherwise, let's check thoroughly that things don't go wrong and there are no leaks whatsoever (I spotted 5 uses of `goto` and admittedly I am scared)

Comment (kostasrim) [reply src/server/stream_family.cc:1589]: @dranikpg do you know which PR?

**Review by kostasrim — CHANGES_REQUESTED** [review]
Good work :muscle: Try to clean up the initializations either by adding constructors or list initialize (I haven't commented on each of the occurrences, so plz take a look on the ones I missed). 

I will pull your branch and play around to check a little bit more the implementation + the goto staff. 

All in all, nice work, tough PR :fire:

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1554]: I had doubts about whether I should use `SendError` here. Thanks!

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:96]: Yeah, you're right. They are exclusive. Let me see if I can optimize that part.

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1589]: True, I also avoid using `goto`s. But didn't find any better way to do this

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1891]: Agreed, it is even better.

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1572]: Sure

File: src/server/stream_family.cc
Comment (dranikpg) [reply src/server/stream_family.cc:1589]: It's #1425, its was on hold because it reduced the likelihood of catching blocking bugs

File: src/server/stream_family.cc
Comment (romange) [reply src/server/stream_family.cc:1561]: To follow up:

returning a status is relevant only for single-shard transactions like GET/SET etc.

here it's a quite complicated multi-shard transaction so returning a status via return value is not relevant (this lambda runs multiple times). Instead, we use `res_sitems` to track a status per shard. It should be   `vector<OpResult<SomeValue>>` where `SomeValue` in this case is `unordered_map<...>`.

you also should not use `stop` variable here because it will be written concurrently by multiple threads. Instead, when the first home (execute call) finishes, you can observe the status of each result in   `res_sitems` and act accordingly.

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1575]: each shard may return a potantially differrent error OpStatus, and you may return the error in the coordinator thread based on the first error you stumble upon.

File: src/server/stream_family.cc
Comment (romange) [reply src/server/stream_family.cc:1589]: I suggest that you duplicate the code here, and just run the callback and return early.

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1606]: Do you have to parse AFTER you start running a transaction?
It's much better to parse everything before you start with the transaction and return errors earlier. This way you know that you continue with consistent arguments. Also If you do this you could factor out this code into a helper function that will return failure/success and this way you won't need "goto".

``` 
if (!ParseArgs(args, cntx)) {
  return;
}
```

Comment (Abhra303) [reply src/server/stream_family.cc:1590]: Yeah, I was able to get rid of `goto` in my latest push. It is cleaner. Please take a look at it.

**Comment by Abhra303** [comment]
I addressed all the reviews in the newest push. Let me know if I missed something.

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:940]: `const OpArgs&` and `const GroupConsumerPairOpts&`

Comment (romange) [inline src/server/stream_family.cc:940]: I do not see that this function returns any status besides the value. If it can not fail, it's better remove `OpResult` wrapping.

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1600]: `"Same stream specified multiple time"` appears multiple times (the irony).
Please factor out into a constant and use it everywhere.

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1861]: I do not understand this condition. When does it happen that `opts->read_group` is false? 
what's the expected behavior in that case?

Comment (Abhra303) [reply src/server/stream_family.cc:1861]: In case of `xread` group command.

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1890]: Probably not the best name for the function.  I see it's being called only here. What does it do? Please add description above the function definition.

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:940]: Ok, sure

**Review by romange — CHANGES_REQUESTED** [review]
Looks indeed better. I gave some comments.

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1890]: `XReadGeneric2` is a generic helper function for `XReadGeneric`. I named it like the way `OpSetId2` is. Will add comments :)

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1873]: either use `res_pairs[sid] = move(items_res)` or just assign directly:
`res_pairs[sid] = OpGetGroupConsumerPairs(...)`

File: src/server/stream_family.cc
Comment (romange) [reply src/server/stream_family.cc:1865]: I prefer that this logic will be done earlier in the flow, i.e. it seems you do not need to schedule this hop at all for XREAD. Instead of checking inside shard callback, 
you can probably run it like this after the schedule call.

```
if (opts->read_group) {
  cntx->transaction->Execute(std::move(cb), false);
}
```

File: src/server/stream_family.cc
Comment (romange) [inline src/server/stream_family.cc:1880]: it's relevant only if  `opts->read_group` is set?

Comment (Abhra303) [reply src/server/stream_family.cc:1880]: yeah, in case of `xread`, the `res_pairs[i].size()` = 0

File: src/server/stream_family_test.cc
Comment (kostasrim) [inline src/server/stream_family_test.cc:162]: I would to see some more test cases, specifically the following queries fail (I pulled your branch locally and played around with redis-ci):

Test dataset:
```
Create a new stream:
xadd mystream * id 1
>> "1688466177147-0"
xadd mystream * id 2
>> "1688466178855-0"
xadd mystream * id 3
>> "1688466180281-0"

bgsave 
>> OK //Store the snapshot so you can reload the streams on crashes and reproduce the bugs below easier

And now let's test print with xread:

xread STREAMS mystream 1688466177146-0

1) 1) "mystream"
2) 1) 1) "1688466177147-0"
        2) 1) "id"
            2) "1"
    2) 1) "1688466178855-0"
        2) 1) "id"
            2) "2"
    3) 1) "1688466180281-0"
        2) 1) "id"
            2) "3"
```

checks without creating a consumer group via `xgroup create`    
``` 
1. xreadgroup GROUP group_a client_1 NOACK STREAMS mystream > 
    result -> dragonfly segfaults
2. xreadgroup GROUP group_a client_1 COUNT 10 BLOCK 10000 NOACK STREAMS mystream >
    result -> dragonfly segfaults
3. xreadgroup GROUP group_a client_1 STREAMS mystream > 
    result -> dragonfly sefaults
4. xreadgroup GROUP group_a client_1 STREAMS mystream 1688467144385-0 
    result -> NIL (prints an error on redis that consumer group is not created)
```

File: src/server/stream_family_test.cc
Comment (kostasrim) [inline src/server/stream_family_test.cc:219]: :100:

File: src/server/stream_family.cc
Comment (kostasrim) [reply src/server/stream_family.cc:2028]: am I missing something?

File: src/server/stream_family.cc
Comment (kostasrim) [inline src/server/stream_family.cc:1533]:
```suggestion
    id_indx++;
```

File: src/server/stream_family.cc
Comment (kostasrim) [inline src/server/stream_family.cc:1768]: We can simplify here, see suggestion :)

**Comment by kostasrim** [comment]
+1000 for removing the `goto` <3

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:2028]: The arity is the number of arguments the command requires atleast. The required number of arguments (atleast) for xreadgroup is indeed 6. That's why I added `-6` here. I am not sure how redis count it. But if you see the `xread`, you will see that the redis json has arity `-4` for `xread` command.

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1983]: Strange, I thought I had removed those already

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:1865]: sounds good, thanks

File: src/server/stream_family.cc
Comment (Abhra303) [reply src/server/stream_family.cc:730]: This is ok because in each iteration, `OpRange` gives a new `Recordvec` containing  a new `Record`. So on each iteration, it would move distinct `Record` to `result`.

File: src/server/stream_family_test.cc
Comment (Abhra303) [reply src/server/stream_family_test.cc:162]: added test cases to address the "group doesn't exist" situations. The error is now fixed

File: src/server/stream_family.cc
Comment (romange) [reply src/server/stream_family.cc:1876]: in that case lets move it into the block `if (opts->read_group) {` above

**Review by romange — DISMISSED** [review]
LGTM unless I am missing anything here.
@kostasrim  are there any pending issues here?

**Review by kostasrim — APPROVED** [review]
Looks good :man_cook: 

:ship: :it:

**Comment by Abhra303** [comment]
@romange  All are done :)

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]

**Pull request merged**
